Add py-mlflow and its dependencies

heerener opened the pull request.

heerener committed:
Add py-mlflow and its dependencies

heerener committed:
mlflow: fix syntax error in package.py

heerener committed:
Merge branch 'develop' into add-py-mlflow

tldahlgren commented on var/spack/repos/builtin/packages/py-databricks-cli/package.py:
```diff
@@ -0,0 +1,24 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/py-gunicorn/package.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/py-prometheus-flask-exporter/package.py:
```diff
@@ -0,0 +1,20 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/py-querystring-parser/package.py:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
```
```suggestion
# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```

tldahlgren commented on var/spack/repos/builtin/packages/py-databricks-cli/package.py:
```diff
@@ -0,0 +1,24 @@
+    pypi = 'databricks-cli/databricks_cli-0.17.4.tar.gz'
+
+    version('0.17.4', sha256='bc0c4dd082f033cb6d7978cacaca5261698efe3a4c70f52f98762c38db925ce0')
+    version('0.14.3', sha256='bdf89a3917a3f8f8b99163e38d40e66dc478c7408954747f145cd09816b05e2c')
```
Is this intentionally `0.14.3` or did you mean it to be `0.17.3`?

tldahlgren commented on var/spack/repos/builtin/packages/py-databricks-cli/package.py:
```diff
@@ -0,0 +1,24 @@
+    """A command line interface for Databricks."""
+
+    homepage = "https://pypi.org/project/databricks-cli/"
+    pypi = 'databricks-cli/databricks_cli-0.17.4.tar.gz'
```
Looks like source distributions use dashes, not underscores.  At least that is the case for 0.17.4.

tldahlgren replied:
Making that correction, I *can* confirm the sha256 for `0.17.4` and `0.14.3`.

tldahlgren reviewed (changes requested):
The ending year for the copyright for each package needs to be corrected (`2020`->`2022`).  More importantly, there is a typo for `py-databricks-cli` pypi value.

tldahlgren commented:
You'll also need to replace single quotes with double to pass the `black` checks.

adamjstewart commented on var/spack/repos/builtin/packages/py-databricks-cli/package.py:
```diff
@@ -0,0 +1,24 @@
+    version('0.17.4', sha256='bc0c4dd082f033cb6d7978cacaca5261698efe3a4c70f52f98762c38db925ce0')
+    version('0.14.3', sha256='bdf89a3917a3f8f8b99163e38d40e66dc478c7408954747f145cd09816b05e2c')
+
+    depends_on('python@3.6:', type=('build', 'run'))
```
This looks wrong

heerener replied:
According to the [pypi page](https://pypi.org/project/databricks-cli/) it should be all right:

> Requirements
>
>     Python Version > 2.7.9 or > 3.6

adamjstewart commented on var/spack/repos/builtin/packages/py-databricks-cli/package.py:
```diff
@@ -0,0 +1,24 @@
+    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+
+    depends_on('py-click@6.7:', type=('run'))
+    depends_on('py-requests@2.17.3:', type=('run'))
+    depends_on('py-tabulate@0.7.7:', type=('run'))
+    depends_on('py-six@1.10.0:', type=('run'))
```
Missing deps, wrong versions:
```
    install_requires=[                                                                   
        # Note: please keep this in sync with `requirements.txt`.                        
        'click>=7.0',                                                                    
        'pyjwt>=1.7.0',                                                                  
        'oauthlib>=3.1.0',                                                               
        'requests>=2.17.3',                                                              
        'tabulate>=0.7.7',                                                               
        'six>=1.10.0',                                                                   
        'configparser>=0.3.5;python_version < "3.6"',                                    
    ],         
```

adamjstewart commented on var/spack/repos/builtin/packages/py-gunicorn/package.py:
```diff
@@ -0,0 +1,18 @@
+    version('20.1.0', sha256='e0a968b5ba15f8a328fdfd7ab1fcb5af4470c28aaf7e55df02a99bc13138e6e8')
+
+    depends_on('python@3.5:', type=('build', 'run'))
+    depends_on('py-setuptools@3.0:', type='build')
```
```suggestion
    depends_on('py-setuptools@3.0:', type=('build', 'run'))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,39 @@
+    version('2.0.1', sha256='7ce6caf3c6acb022d6f5ce8a0995a92be1db524ae16aade1f83da661cdf993de')
+    version('1.17.0', sha256='4898c58899e3101e09e2b37cf5bee7db04c5d73389a56942d3ef5a5e4396799e')
+
+    depends_on('python@3.6:', type=('build', 'run'))
```
It's 3.8, at least for 2.0.1, didn't check 1.17.0

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,39 @@
+    depends_on('py-prometheus-flask-exporter', type=('run'))
+    depends_on('py-querystring-parser', type=('run'))
+    depends_on('py-sqlparse@0.3.1:', type=('run'))
+    depends_on('py-sqlalchemy', type=('run'))
```
Seems to be missing some deps:
```
alembic<2
docker<7,>=4.0.0
Flask<3
numpy<2
scipy<2
pandas<2
querystring_parser<2
sqlalchemy<2,>=1.4.0
gunicorn<21; platform_system != 'Windows'
waitress<3; platform_system == 'Windows'
scikit-learn<2
pyarrow<11,>=4.0.0
shap<1,>=0.40
markdown<4,>=3.3
Jinja2<4,>=2.11; platform_system != 'Windows'
Jinja2<4,>=3.0; platform_system == 'Windows'
matplotlib<4
```
```
click<9,>=7.0
cloudpickle<3
databricks-cli<1,>=0.8.7
entrypoints<1
gitpython<4,>=2.1.0
pyyaml<7,>=5.1
protobuf<5,>=3.12.0
pytz<2023
requests<3,>=2.17.3
packaging<22
importlib_metadata<6,>=3.7.0,!=4.7.0
sqlparse<1,>=0.4.0
```

adamjstewart commented on var/spack/repos/builtin/packages/py-prometheus-flask-exporter/package.py:
```diff
@@ -0,0 +1,20 @@
+    version('0.18.2', sha256='fc487e385d95cb5efd045d6a315c4ecf68c42661e7bfde0526af75ed3c4f8c1b')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-prometheus-client', type=('run'))
+    depends_on('py-flask', type=('run'))
```
```suggestion
    depends_on('py-prometheus-client', type=('build', 'run'))
    depends_on('py-flask', type=('build', 'run'))
```
Needed by import tests

adamjstewart commented on var/spack/repos/builtin/packages/py-querystring-parser/package.py:
```diff
@@ -0,0 +1,18 @@
+
+    version('1.2.4', sha256='644fce1cffe0530453b43a83a38094dbe422ccba8c9b2f2a1c00280e14ca8a62')
+
+    depends_on('py-six', type=('run'))
```
```suggestion
    depends_on('py-six', type=('build', 'run'))
```

adamjstewart reviewed (changes requested):
Style tests (single quotes instead of double quotes) can be fixed with `spack style --fix -t black` on the command line or `@spackbot fix style` on this PR. There's also a way to automatically update the licenses but I don't remember it.

heerener committed:
py-mlflow: cleanup

Process review remarks, add missing dependencies, add skinny variant

heerener commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,59 @@
+    depends_on("py-pandas@:1", type=("build", "run"), when="~skinny")
+    depends_on("py-querystring-parser@:1", type=("build", "run"), when="~skinny")
+    depends_on("py-sqlalchemy@1.4.0:1", type=("build", "run"), when="~skinny")
+    depends_on("py-gunicorn@:20", type=("build", "run"), when="~skinny")
```
Is there a way to specify this dependency (same for py-jinja2@2.11:3) for "any OS except Windows"?

adamjstewart replied:
There is no "not" in the spec syntax, the best you can do at the moment is:
```python
for platform in ["linux", "darwin", "cray"]:
    depends_on("py-gunicorn@:20", when="~skinny platform=" + platform, type=("build", "run"))
```

heerener committed:
Apply suggestions from code review

Co-authored-by: Tamara Dahlgren <35777542+tldahlgren@users.noreply.github.com>
Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

heerener commented:
@spackbot fix style

heerener committed:
[@spackbot] updating style on behalf of heerener

heerener committed:
Fix flake8 issues

heerener committed:
More formatting fixes

heerener committed:
Fix py-waitress dependency version

heerener committed:
py-mlflow: platform-specific dependency

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,62 @@
+    version("2.0.1", sha256="7ce6caf3c6acb022d6f5ce8a0995a92be1db524ae16aade1f83da661cdf993de")
+    version("1.17.0", sha256="4898c58899e3101e09e2b37cf5bee7db04c5d73389a56942d3ef5a5e4396799e")
+
+    variant(
+        "skinny",
+        default=False,
+        description="Lightweight MLflow package without SQL storage, server, UI, "
+        "or data science dependencies.",
+    )
```
I think we should actually remove this variant, and here's why.

I usually try to emulate pip as best I can so that installing Python packages with Spack is no different than installing them with pip. Pip has a concept of "extras" which is basically the same as our "variants", so it's not uncommon to add variants to match those. But "skinny" isn't an extra, it's an entirely different package: https://pypi.org/project/mlflow-skinny/. The source code for both is the same, but you don't `pip install mlflow[skinny]`, you `pip install mlflow-skinny`. So I think if you want `mlflow-skinny`, there should be a new `py-mlflow-skinny` package for it.

Although `mlflow-skinny` only requires the skinny requirements, `mlflow` proper requires both core and skinny requirements. So you can remove the variant but keep all the deps.

heerener replied:
Got it - thanks for the clear explanation!

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,62 @@
+    depends_on("py-sqlalchemy@1.4.0:1", type=("build", "run"), when="~skinny")
+    for platform in ["linux", "darwin", "cray"]:
+        depends_on("py-gunicorn@:20", type=("build", "run"), when=f"~skinny platform={platform}")
+    depends_on("py-waitress@:2", type=("build", "run"), when="os=windows ~skinny")
+    depends_on("py-scikit-learn@:1", type=("build", "run"), when="~skinny")
+    depends_on("py-pyarrow@4.0.0:10", type=("build", "run"), when="~skinny")
+    depends_on("py-shap@0.40:0", type=("build", "run"), when="~skinny")
+    depends_on("py-markdown@3.3:3", type=("build", "run"), when="~skinny")
+    for platform in ["linux", "darwin", "cray"]:
+        depends_on("py-jinja2@2.11:3", type=("build", "run"), when=f"~skinny platform={platform}")
+    depends_on("py-jinja2@3.0:3", type=("build", "run"), when="os=windows ~skinny")
```
```suggestion
    depends_on("py-waitress@:2", type=("build", "run"), when="platform=windows ~skinny")
    depends_on("py-scikit-learn@:1", type=("build", "run"), when="~skinny")
    depends_on("py-pyarrow@4.0.0:10", type=("build", "run"), when="~skinny")
    depends_on("py-shap@0.40:0", type=("build", "run"), when="~skinny")
    depends_on("py-markdown@3.3:3", type=("build", "run"), when="~skinny")
    for platform in ["linux", "darwin", "cray"]:
        depends_on("py-jinja2@2.11:3", type=("build", "run"), when=f"~skinny platform={platform}")
    depends_on("py-jinja2@3.0:3", type=("build", "run"), when="platform=windows ~skinny")
```
Windows is a platform just like linux/macos/cray

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,62 @@
+        depends_on("py-jinja2@2.11:3", type=("build", "run"), when=f"~skinny platform={platform}")
+    depends_on("py-jinja2@3.0:3", type=("build", "run"), when="os=windows ~skinny")
+    depends_on("py-matplotlib@:3", type=("build", "run"), when="~skinny")
+    depends_on("py-prometheus-flask-exporter", type=("build", "run"), when="~skinny")
```
I don't see this dep documented, at least in 2.0.1

heerener replied:
I still saw it being imported so I kept it, but it seems to be part of [extras_require](https://github.com/mlflow/mlflow/blob/master/setup.py#L152).
I'll remove it for now, and add it if I ever get around to adding the extras variant.

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,62 @@
+    depends_on("python@3.8:", type=("build", "run"), when="@2.0.1:")
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-click@7.0:", type=("build", "run"))
+    depends_on("py-cloudpickle", type=("build", "run"))
+    depends_on("py-databricks-cli@0.8.7:", type=("build", "run"))
+    depends_on("py-entrypoints", type=("build", "run"))
+    depends_on("py-gitpython@2.1.0:", type=("build", "run"))
+    depends_on("py-pyyaml", type=("build", "run"))
+    depends_on("py-protobuf@3.6.0:", type=("build", "run"))
+    depends_on("py-pytz", type=("build", "run"))
+    depends_on("py-requests@2.17.3:", type=("build", "run"))
```
These are missing their upper bounds

adamjstewart commented on var/spack/repos/builtin/packages/py-mlflow/package.py:
```diff
@@ -0,0 +1,62 @@
+    depends_on("py-pytz", type=("build", "run"))
+    depends_on("py-requests@2.17.3:", type=("build", "run"))
+    depends_on("py-packaging@:21", type=("build", "run"))
+    depends_on("py-importlib-metadata@3.7.0:4.6.4,4.7.1:5", type=("build", "run"))
```
```suggestion
    depends_on("py-importlib-metadata@3.7:4.6,4.7.1:5", type=("build", "run"))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-pyarrow/package.py:
```diff
@@ -15,6 +15,7 @@ class PyPyarrow(PythonPackage, CudaPackage):
+    version("10.0.1", sha256="1a14f57a5f472ce8234f2964cd5184cccaa8df7e04568c64edc33b23eb285dd5")
     version("8.0.0", sha256="4a18a211ed888f1ac0b0ebcb99e2d9a3e913a481120ee9b1fe33d3fedb945d4e")
     version("7.0.0", sha256="da656cad3c23a2ebb6a307ab01d35fce22f7850059cffafcb90d12590f8f4f38")
     version("4.0.1", sha256="11517f0b4f4acbab0c37c674b4d1aad3c3dfea0f6b1bb322e921555258101ab3")
```
Setuptools version needs to be updated, six can actually be removed since we don't have that version

adamjstewart replied:
Need to update for-loop below for arrow dep

adamjstewart commented on var/spack/repos/builtin/packages/py-shap/package.py:
```diff
@@ -0,0 +1,16 @@
+    homepage = "https://github.com/slundberg/shap/"
+    pypi = "shap/shap-0.41.0.tar.gz"
+
+    version("0.41.0", sha256="a49ea4d65aadbc845a695fa3d7ea0bdfc8c928b8e213b0feedf5868ade4b3ca5")
```
Missing all of its deps. Build dep on setuptools +:
```
        install_requires=['numpy', 'scipy', 'scikit-learn', 'pandas', 'tqdm>4.25.0',     
                          'packaging>20.9', 'slicer==0.0.7', 'numba', 'cloudpickle'],  
```

adamjstewart commented on var/spack/repos/builtin/packages/py-waitress/package.py:
```diff
@@ -0,0 +1,18 @@
+    homepage = "https://github.com/Pylons/waitress/"
+    pypi = "waitress/waitress-2.1.2.tar.gz"
+
+    version(
+        "2.1.2",
+        sha256="780a4082c5fbc0fde6a2fcfe5e26e6efc1e8f425730863c04085769781f51eba",
+    )
```
Can you put this on one line?

adamjstewart replied:
Missing build dep:
```
requires = ["setuptools >= 41"]
```

heerener committed:
Update var/spack/repos/builtin/packages/py-mlflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

heerener committed:
Update var/spack/repos/builtin/packages/py-mlflow/package.py

Co-authored-by: Adam J. Stewart <ajstewart426@gmail.com>

heerener committed:
Process review remarks

heerener committed:
Fix typo in dependency version

adamjstewart commented on var/spack/repos/builtin/packages/py-pyarrow/package.py:
```diff
@@ -34,14 +35,13 @@ class PyPyarrow(PythonPackage, CudaPackage):
     depends_on("python@3.7:", type=("build", "run"), when="@7.0.0:")
     depends_on("py-setuptools", type="build")
-    depends_on("py-setuptools@38.6.0:", type="build", when="@7.0.0:")
+    depends_on("py-setuptools@65.5.0:", type="build", when="@7.0.0:")
```
This should only be changed for the latest release, 7 still depends on 38.6+ right?

adamjstewart commented on var/spack/repos/builtin/packages/py-shap/package.py:
```diff
@@ -0,0 +1,28 @@
+    depends_on("py-scipy", type=("build", "run"))
+    depends_on("py-scikit-learn", type=("build", "run"))
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("py-tqdm@4.26:", type=("build", "run"))
```
```suggestion
    depends_on("py-tqdm@4.25.1:", type=("build", "run"))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-shap/package.py:
```diff
@@ -0,0 +1,28 @@
+    depends_on("py-scikit-learn", type=("build", "run"))
+    depends_on("py-pandas", type=("build", "run"))
+    depends_on("py-tqdm@4.26:", type=("build", "run"))
+    depends_on("packaging@20.10", type=("build", "run"))
```
```suggestion
    depends_on("packaging@20.10:", type=("build", "run"))
```

adamjstewart commented on var/spack/repos/builtin/packages/py-slicer/package.py:
```diff
@@ -0,0 +1,18 @@
+
+    version("0.0.7", sha256="f5d5f7b45f98d155b9c0ba6554fa9770c6b26d5793a3e77a1030fb56910ebeec")
+
+    depends_on("py-setuptools@41:", type="build")
```
Where does 41+ come from?

heerener committed:
py-shap: fix dependencies

heerener committed:
py-arrow: fix dependencies

heerener committed:
py-slicer: remove py-setuptools explicit version

matz-e committed:
py-pyarrow: dataset variant and pass options through environment

It appears there are some issues when using `pip install` instead of
`python setup.py` - this setup_build_environment should fix that.

heerener commented on var/spack/repos/builtin/packages/py-pyarrow/package.py:
```diff
@@ -62,6 +65,13 @@ class PyPyarrow(PythonPackage, CudaPackage):
 
     patch("for_aarch64.patch", when="@0 target=aarch64:")
 
+    def setup_build_environment(self, env):
```
https://github.com/spack/spack/issues/35160#issuecomment-1406463878 for context / info

adamjstewart replied:
It might be better to decouple this from `install_options`. These flags are only for pip, I'm planning on adding support for build/installer in the future which will likely have different flags.

thomas-bouvier commented:
For reference, I was working on adding package `py-pyarrow@10.0.1` there https://github.com/spack/spack/pull/35068 but stumbled upon difficulties related to the `+parquet` variant.

adamjstewart commented:
Need to resolve merge conflicts

heerener committed:
Merge branch 'develop' into add-py-mlflow

heerener committed:
py-pyarrow: review remark

Decouple setup_build_environment from install_options

heerener committed:
py-pyarrow: style

heerener committed:
Bump licenses to 2023

heerener commented:
I noticed that the licenses were still set to 2022.
@tldahlgren , github tells me you still have one open change request, but I'm not sure what it is / how to mark it as resolved.

thomas-bouvier commented:
@spack run pipeline

tldahlgren reviewed (approved):
Re-confirmed all affected version sha256.

tldahlgren commented:
> I noticed that the licenses were still set to 2022.
> @tldahlgren , github tells me you still have one open change request, but I'm not sure what it is / how to mark it as resolved.

Last I heard Spack may be getting rid of the copyright year so this doesn't have to be changed every year.

Looks like @adamjstewart addressed the change request issue and has since approved the PR.
